staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go

```diff
@@ -32,24 +33,28 @@ const (
 // ready is a three state condition variable that blocks until is Ready if is not Stopped.
 // Its initial state is Pending.
 type ready struct {
-	state status
-	c     *sync.Cond
+	state     status
+	lock      sync.Mutex
+	pendingCh chan struct{} // blocks until is ready or stopped
 }
 
 func newReady() *ready {
 	return &ready{
-		c:     sync.NewCond(&sync.RWMutex{}),
-		state: Pending,
+		pendingCh: make(chan struct{}),
+		state:     Pending,
 	}
 }
 
 // wait blocks until it is Ready or Stopped, it returns an error if is Stopped.
-func (r *ready) wait() error {
-	r.c.L.Lock()
-	defer r.c.L.Unlock()
-	for r.state == Pending {
-		r.c.Wait()
+func (r *ready) wait(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-r.pendingCh:
```

Comment:
Currently - this seems to be racy, as you're accessing `r.pendingCh` and `r.pendingCh` can be assigned in line 93.

I like the fact that you don't have additional goroutine that I have, but I'm a bit afraid of this approach - it remind me a bit the journey that we had on:
https://github.com/kubernetes/kubernetes/pull/108414#discussion_r818374157

and ended up finally not using channels...

Reply:
that approach was more complicated because was using channels only, this uses a mutex and the enum that makes it easier to reason about than the other.

The access to the channel are always done through a `select`statement, and the delete/create operations are locked, I assume that the select works for handling the concurrency on the reads

Reply:
FWICT, `select` makes the operation safe only in terms of the channels.
But you're actually allowing the the actual channel to be changed.

So race I have on my mind is:
- I'm entering `wait()` and I'm now entering the `select`
- at exactly the same time `set(false)` is called, it takes the lock and is running:
```
r.pendingCh = make(chan struct{})
```

So now the read/write race is not on the channel itself, but at the level of accessing the field of ready type (its pendingCh field).

I'm fairly sure it's not safe.

Reply:
yeah, you are absolutely right, exactly as you say
```
func Test_newReadyRacy(t *testing.T) {
	ready := newReady()
	ready.set(false)
	// create 10 goroutines waiting for ready
	for i := 0; i < 1000; i++ {
		go func() {
			ready.wait(context.Background())
		}()
		go func() {
			ready.set(false)
		}()
		go func() {
			ready.set(true)
		}()
	}
	ready.stop()
}
```

```
 go test -run Test_newReadyRacy -v -race -count 1 k8s.io/apiserver/pkg/storage/cacher/
=== RUN   Test_newReadyRacy
==================
WARNING: DATA RACE
Write at 0x00c000306f28 by goroutine 78:
  k8s.io/apiserver/pkg/storage/cacher.(*ready).set()
      /usr/local/google/home/aojea/src/hotfix/vendor/k8s.io/apiserver/pkg/storage/cacher/ready.go:93 +0x18b
  k8s.io/apiserver/pkg/storage/cacher.Test_newReadyRacy.func2()
      /usr/local/google/home/aojea/src/hotfix/vendor/k8s.io/apiserver/pkg/storage/cacher/ready_test.go:57 +0x32

Previous read at 0x00c000306f28 by goroutine 86:
  k8s.io/apiserver/pkg/storage/cacher.(*ready).wait()
      /usr/local/google/home/aojea/src/hotfix/vendor/k8s.io/apiserver/pkg/storage/cacher/ready.go:53 +0x87
  k8s.io/apiserver/pkg/storage/cacher.Test_newReadyRacy.func1()
      /usr/local/google/home/aojea/src/hotfix/vendor/k8s.io/apiserver/pkg/storage/cacher/ready_test.go:54 +0x4a

Goroutine 78 (running) created at:
  k8s.io/apiserver/pkg/storage/cacher.Test_newReadyRacy()
      /usr/local/google/home/aojea/src/hotfix/vendor/k8s.io/apiserver/pkg/storage/cacher/ready_test.go:56 +0x1c4
  testing.tRunner()
      /usr/local/google/home/aojea/sdk/go1.19.4/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/google/home/aojea/sdk/go1.19.4/src/testing/testing.go:1493 +0x47

Goroutine 86 (running) created at:
  k8s.io/apiserver/pkg/storage/cacher.Test_newReadyRacy()
      /usr/local/google/home/aojea/src/hotfix/vendor/k8s.io/apiserver/pkg/storage/cacher/ready_test.go:53 +0x15a
  testing.tRunner()
      /usr/local/google/home/aojea/sdk/go1.19.4/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/google/home/aojea/sdk/go1.19.4/src/testing/testing.go:1493 +0x47
==================
    testing.go:1319: race detected during execution of test
--- FAIL: Test_newReadyRacy (0.04s)
=== CONT  
    testing.go:1319: race detected during execution of test
FAIL
FAIL    k8s.io/apiserver/pkg/storage/cacher     0.154s
```

Reply:
that implies a transition from ready (no channel) to pending (create channel) though, that we don't use ... if we simplify the state machine to be unidirectional things are simpler

pending -> started -> stopped
               |-----------> stopped

Reply:
@wojtek-t fixed in the last push

```
go test  -v -race -count 1 k8s.io/apiserver/pkg/storage/cacher/ -c
aojea@aojea:~/src/hotfix$ stress ./cacher.test -test.run Test_newReadyRacy
5s: 1217 runs so far, 0 failures
10s: 2491 runs so far, 0 failures
15s: 3797 runs so far, 0 failures
20s: 5069 runs so far, 0 failures
```